### PR TITLE
5 api/articles endpoint added 

### PR DESCRIPTION
**Pull request by cls-c**
Completed task 5: 
- Added api/articles endpoint 
- allow and accept sort_by 
- fixed test based on comment on task 4.

### CODE REVIEW

**Comment by cls-c** [comment]
Please do not review this PR: 
For task 5 please review : https://github.com/cls-c/nc-news/pull/8

File: __tests__/app.test.js [hunk 1]
```diff
@@ -96,13 +93,91 @@ describe("GET /api/articles/:article_id", () => {
...
+      });
+  });
+  test("should return  404 badRequest if provided id is non existent", () => {
+    const articleId = 9999;
     return request(app)
       .get(`/api/articles/${articleId}`)
       .expect(404)
       .then((response) => {
-        expect(response.body.msg).toBe("Bad Request: ID provided has not been found.");
+        expect(response.body.msg).toBe(
```
Comment (nc-jodie-cheung) [inline __tests__/app.test.js:113]: This should be  **.get(`/api/articles/9999`**

File: __tests__/app.test.js [hunk 2]
```diff
@@ -96,13 +93,91 @@ describe("GET /api/articles/:article_id", () => {
         );
       });
   });
-  test("should return  404 badRequest if provided id is invalid", () => {
+  test("should return  400 badRequest if provided id is invalid", () => {
     const articleId = "invalid";
+    return request(app)
+      .get(`/api/articles/${articleId}`)
+      .expect(400)
```
Comment (nc-jodie-cheung) [inline __tests__/app.test.js:100]: Change this to  **.get(`/api/articles/notanumber`**

File: __tests__/app.test.js [hunk 3]
```diff
@@ -96,13 +93,91 @@ describe("GET /api/articles/:article_id", () => {
...
+        expect(body.article).toBeSorted({ key: sortingKey });
+      });
+  });
+
+  test("should return  404 badRequest if sort_by parameter is invalid", () => {
+    const sortingKey = "invalid";
+    return request(app)
+      .get(`/api/articles?sort_by=${sortingKey}`)
+      .expect(404)
+      .then((response) => {
```
Comment (nc-jodie-cheung) [inline __tests__/app.test.js:177]: Same as comments above, you can stick to this way but I'd suggest that you can just change line 175 to:  **.get(`/api/articles?sort_by=invalid**

File: __tests__/app.test.js [hunk 4]
```diff
@@ -96,13 +93,91 @@ describe("GET /api/articles/:article_id", () => {
...
+          ])
+        );
+      });
+  });
+  test("should return an array of objects, sorted by created_at date by default  ", () => {
+    return request(app)
+      .get("/api/articles")
+      .expect(200)
+      .then(({ body }) => {
+        expect(body.article).toBeSorted({ key: "created_at" });
```
Comment (nc-jodie-cheung) [inline __tests__/app.test.js:159]: Nice use of the jest sorted library here 👍

File: model.js [hunk 5]
```diff
@@ -10,16 +10,52 @@ exports.fetchTopics = () => {
...
+};
+exports.fetchArticles = (sortingKey) => {
+  let sortBy = "articles.created_at";
+  const acceptedKey = {
+    author: "articles",
+    title: "articles",
+    article_id: "articles",
+    topic: "articles",
+    votes: "articles",
+    count_comment: "t2",
```
Comment (nc-jodie-cheung) [inline model.js:40]: You're creating an object to get the keys from it, I'd recommend creating an array with the keys to check if they're valid.

File: model.js [hunk 6]
```diff
@@ -10,16 +10,52 @@ exports.fetchTopics = () => {
...
+    }
+
+    sortBy = `${
+      Object.entries(acceptedKey).find((element) => {
+        return element[0] === sortingKey;
+      })[1]
+    }.${sortingKey}`;
+  }
+  const query = format(
+    `SELECT articles.author,articles.title,articles.article_id,articles.topic,articles.created_at,articles.votes,articles.article_img_url, t2.count_comment AS comment_count FROM articles LEFT JOIN (SELECT article_id, COUNT(*) AS count_comment FROM comments GROUP BY article_id) AS t2 ON articles.article_id = t2.article_id ORDER BY ${sortBy} ASC;`
```
Comment (nc-jodie-cheung) [inline model.js:56]: This can be **SELECT articles.***

File: model.js [hunk 7]
```diff
@@ -10,16 +10,52 @@ exports.fetchTopics = () => {
...
+      return Promise.reject({ msg: "Non-existent id" });
     } else {
       return data.rows;
     }
   });
-}
\ No newline at end of file
+};
+exports.fetchArticles = (sortingKey) => {
+  let sortBy = "articles.created_at";
```
Comment (nc-jodie-cheung) [inline model.js:33]: This task doesn't require you to create a query. The sort-by should be set to default. It looks like you have started task 11 here.

File: __tests__/app.test.js [hunk 8]
```diff
@@ -96,13 +93,91 @@ describe("GET /api/articles/:article_id", () => {
...
+      });
+  });
+  test("If specified a sortby optional parameter, will return object that is ordered by sorted by optional parameter, in asending order", () => {
+    const sortingKey = "author";
+    return request(app)
+      .get(`/api/articles?sort_by=${sortingKey}`)
+      .expect(200)
+      .then(({ body }) => {
+        expect(body.article).toBeSorted({ key: sortingKey });
+      });
```
Comment (nc-jodie-cheung) [inline __tests__/app.test.js:169]: You don't need this test, it should work by default. You have already done a test for this above.

**Review by nc-jodie-cheung — COMMENTED** [review]
Hey 👋 please see the comments below, it seems like you have started writing code for task 11. No need to change but please bear the comments in mind!

**Pull request merged**